frontend/lib/parsing/parsing-queries.cpp

```diff
@@ -917,6 +917,124 @@ ConfigSettingsList& configSettings(Context* context) {
   return QUERY_END(result);
 }
 
+const uast::Attributes* idToAttributes(Context* context, ID id) {
+  const uast::Attributes* ret = nullptr;
+  if (id.isEmpty()) return ret;
+
+  auto ast = parsing::idToAst(context, id);
+  if (ast && ast->isDecl()) {
+    auto decl = ast->toDecl();
+    ret = decl->attributes();
+  }
+
+  return ret;
+}
+
+static std::string
+createDefaultDeprecationMessage(Context* context, const NamedDecl* target) {
+  std::string ret = target->name().c_str();
+  ret += " is deprecated";
+  return ret;
+}
+
+static std::string
+createDefaultUnstableMessage(Context* context, const NamedDecl* target) {
+  std::string ret = target->name().c_str();
+  ret += " is unstable";
+  return ret;
+}
+
+static const ErrorBase* const&
+deprecationWarningForIdQuery(Context* context, ID idMention, ID idTarget) {
+  QUERY_BEGIN(deprecationWarningForIdQuery, context, idMention, idTarget);
+  const ErrorBase* ret = nullptr;
+
+  if (idMention.isEmpty() || idTarget.isEmpty()) return QUERY_END(ret);
+
+  auto attributes = parsing::idToAttributes(context, idTarget);
+  if (!attributes) return QUERY_END(ret);
+
+  bool isDeprecated = attributes->hasPragma(PRAGMA_DEPRECATED) ||
```

Comment:
When would one of these conditions be met but not the other?

Reply:
Today I can do

```chapel
pragma "deprecated"
class C{}
```

Which is different syntactically than

```chapel
deprecated
class C {}
```